### PR TITLE
is_fqdn: only match FULLY QUALIFIED domain names

### LINKED ISSUE #66: is_fqdn function also matches partially qualified hostnames
The [`is_fqdn`](https://github.com/quattor/template-library-core/blob/master/pan/types.pan#L431) function in [types.pan](https://github.com/quattor/template-library-core/blob/master/pan/types.pan) will match partially qualified hostnames as well as proper FQDNs.
For instance, it will match `example.org` (which it should) but will also match `example`.
The [regex](https://github.com/quattor/template-library-core/blob/master/pan/types.pan#L436) will match a properly formed hostname component followed by a literal `.` or the end of the string.

This means there is no functionality available to ensure that a value is actually an FQDN as opposed to a hostname.

Furthermore, the [`is_hostname`](https://github.com/quattor/template-library-core/blob/master/pan/types.pan#L448) function seems to [rely](https://github.com/quattor/template-library-core/blob/master/pan/types.pan#L453) on this behaviour (it should match IPs, FQDNs and partially qualified hostnames) and would be (partially) broken by a fix to `is_fqdn`.

### PR DESCRIPTION
**Pull request by jrha**
Fixes #66

Tested and seems to work as desired!

### CODE REVIEW

File: pan/types.pan [hunk 1]
```diff
@@ -433,7 +433,7 @@ function is_fqdn = {
     if (ARGC != 1 || !is_string(ARGV[0]))
         error("usage: is_fqdn(string)");
 
-    return(match(ARGV[0],'^([a-zA-Z\d]([a-zA-Z\d-]{0,253}[a-zA-Z\d])?(\.|$))+$'));
+    return(match(ARGV[0],'(?=^.{4,253}$)(^((?!-)[a-zA-Z0-9-]{1,63}(?<!-)\.)+[a-zA-Z]{2,63}\.?$)'));
```
Comment (stdweird) [inline pan/types.pan]: inspired by http://stackoverflow.com/questions/11809631/fully-qualified-domain-name-validation?

**Comment by stdweird** [comment]
LGTM

File: pan/types.pan (hunk 1 above)
Comment (jrha) [inline pan/types.pan]: Looks the same @gvasilak didn't cite a source.

**Comment by jrha** [comment]
NB: This will probably break `is_hostname` as it stands.

**Comment by stdweird** [comment]
@jrha why would it do that?

**Comment by jrha** [comment]
Hmm, @gvasilak seemed to think it would, but it would actually cause it to behave as documented so...

NB: This also fixes `is_hostname`!

**Comment by stdweird** [comment]
and breaks `is_network_name`, since it checks if `.domain` via `is_hostname(domain)`, good luck verifying TLDs.

**Comment by stdweird** [comment]
also, now that shorthostnames aren't valid hostnames anymore, `is_network_name` also fails on shorthostnames, which is an issue for our ceph config.

**Comment by jrha** [comment]
`is_network_name` now matches TLDs again, according to it's "documentation" it shouldn't match short host names. So do we change the `ncm-ceph` schema or `is_network_name`?

**Comment by stdweird** [comment]
can we issue a deprecated warning to allow short hostnames as part of the backwards incompatibility? and clean it up to conformto the docs in next release? (or is this not worth it?)

**Comment by jrha** [comment]
I guess we can.

File: pan/types.pan [hunk 2]
```diff
@@ -829,6 +829,20 @@ type type_uppercase = string with {
 
 
 @documentation{
+Checks if the argument is a valid top level domain name (e.g. .example)
+}
+function is_top_level_domain = {
+    if (match(ARGV[0], '^(\.)+[a-zA-Z]{2,63}?$')) return(true);
```
Comment (stdweird) [inline pan/types.pan]: simpler:
```pan
return(match(...));
```

File: pan/types.pan [hunk 3]
```diff
@@ -837,24 +851,25 @@ function is_network_name = {
...
+    if (is_hostname(ARGV[0])) return(true);
+
+    if (is_shorthostname(ARGV[0])) {
+        deprecate(0, "Short hostnames are deprecated as valid network_names.");
+        return(true);
     };
 
+    if (is_top_level_domain(ARGV[0])) return(true);
+
     # Not a hostname. Is it a IP/mask?
```
Comment (stdweird) [inline pan/types.pan]: move this to its own function/type combo?

Comment (jrha) [inline pan/types.pan]: What do you mean? I did add `type_top_level_domain`.

Comment (stdweird) [inline pan/types.pan]: i meant this block of code `is_ip_and_mask`? (i know it's not your code and all that, so feel free to ignore)

Comment (jrha) [inline pan/types.pan]: Oh ok, sure.

File: pan/types.pan [hunk 4]
```diff
@@ -829,6 +829,37 @@ type type_uppercase = string with {
...
+    is_top_level_domain(SELF);
+};
+
+
+function is_ipv4_netmask_pair = {
+    pair = split('\/', 1, ARGV[0]);
+    if (length(pair) == 2) {
+        ip = pair[0];
+        netmask = pair[1];
+        if (is_ip(ip) && (is_ip(netmask) || match(netmask, '^(\d+)$'))) {
```
Comment (stdweird) [inline pan/types.pan]: maybe `(match(netmask, '^(\d+)$') && to_long(netmask) >= 0&&  to_long(netmask) <= 32)`

Comment (jrha) [inline pan/types.pan]: This bit conflates netmasks and prefix lengths, perhaps it is worth defining types for them?

Comment (stdweird) [inline pan/types.pan]: sure, seems like a nice addition.

File: pan/types.pan (hunk 1 above)
Comment (ned21) [inline pan/types.pan]: Drop the "return" and just say: ```match(ARGV[0], ...);```

As per PAN compiler doc, return() should be avoided if possible for performance reasons.

File: pan/types.pan [hunk 5]
```diff
@@ -829,6 +829,37 @@ type type_uppercase = string with {
...
 
 @documentation{
+Checks if the argument is a valid top level domain name (e.g. .example)
+}
+function is_top_level_domain = {
+    return(match(ARGV[0], '^(\.)+[a-zA-Z]{2,63}?$'));
+};
+
+
+type type_top_level_domain = string with {
```
Comment (ned21) [inline pan/types.pan]: Why the separate function call? 

```
type type_top_level_domain = string with match(...); 
```

File: pan/types.pan [hunk 6]
```diff
@@ -829,6 +829,37 @@ type type_uppercase = string with {
...
+function is_ipv4_netmask_pair = {
+    pair = split('\/', 1, ARGV[0]);
+    if (length(pair) == 2) {
+        ip = pair[0];
+        netmask = pair[1];
+        if (is_ip(ip) && (is_ip(netmask) || match(netmask, '^(\d+)$'))) {
+            return(true);
+        };
+    };
+    return(false);
```
Comment (ned21) [inline pan/types.pan]: s/return(false)/false/ for performance

File: pan/types.pan [hunk 7]
```diff
@@ -837,24 +868,20 @@ function is_network_name = {
...
-    };
-    # Not IP/mask: is it .domain?
-    if (index ('.', hst) == 0) {
-        return (is_hostname (substr (hst, 1)));
-    };
+    is_ipv4_netmask_pair(ARGV[0]);
+
     # Everything failed!
-    return (false);
+    return(false);
```
Comment (ned21) [inline pan/types.pan]: Drop return() can just say "false"

File: pan/types.pan (hunk 5 above)
Comment (jrha) [inline pan/types.pan]: The function is for use in `is_network_address`, the type is a freebie.

File: pan/types.pan (hunk 6 above)
Comment (jrha) [inline pan/types.pan]: Neat, didn't know that.

File: pan/types.pan (hunk 4 above)
Comment (stdweird) [inline pan/types.pan]: @jrha if you have no time for this atm, simply open an issue for it so this can be merged.

**Comment by jrha** [comment]
@stdweird: updated

File: pan/types.pan [hunk 8]
```diff
@@ -829,6 +829,47 @@ type type_uppercase = string with {
 
 
 @documentation{
+Checks if the argument is a valid top level domain name (e.g. .example)
+}
+function is_top_level_domain = {
+    return(match(ARGV[0], '^(\.)+[a-zA-Z]{2,63}?$'));
```
Comment (stdweird) [inline pan/types.pan]: remove the `return`

**Comment by stdweird** [comment]
@jrha one more tiny remark

**Comment by stdweird** [comment]
@jrha more LGTM than ever

**Pull request merged**